src/inmanta/agent/executor.py

```diff
@@ -423,8 +417,16 @@ async def clean_environments(self) -> None:
                 # The Virtual Environment could be incomplete or broken
                 environments_on_disk_to_clean.add(absolute_path_file)
 
+        loop = asyncio.get_running_loop()
         for path_env_to_clean in environments_on_disk_to_clean:
-            shutil.rmtree(path_env_to_clean)
+            # If the path to clean is part of the `_environment_map` then we can use the threadpool of the executor to clean
+            # the virtual environment
+            if path_env_to_clean in reverse_environment_map:
+                blueprint = reverse_environment_map[path_env_to_clean]
+                async with self._locks.get(blueprint.blueprint_hash()):
+                    await loop.run_in_executor(self._environment_map[blueprint].thread_pool, shutil.rmtree, path_env_to_clean)
+            else:
+                shutil.rmtree(path_env_to_clean)
```

Comment:
The more import question is, why is it OK to block the IOloop if it isn't?

and two what makes this race-free with creation?

Reply:
About 2. Isn't the lock enough? We rely on it when we create the environment and if the environment exists in memory in the cleanup, we use the same lock. Or am I missing something?

Reply:
this code is not below the lock